### ISSUE TITLE
Confusion about training objective

### ISSUE THREAD

**Issue by ChengkaiYang** [roggirg/AutoBots#21]
Thanks for your excellent work!
In the following objective, the approximating posterior of the latent variable is p_old(z|y,x1:t).The paper has said it can be calculated because the latent variable z is descrete. p_old(z | y,x1:t) = p_old(z | x1:t) * p_old(y | z,x1:t) / p_old(y,x1:t).When calculating the prior p_old(z | x1:t),the origin code wrote this:"priors = modes_pred.detach().cpu().numpy()",but when calculating p_old(y | z,x1:t),I wonder why don't add detach() operator?
<img width="608" alt="image" src="https://github.com/roggirg/AutoBots/assets/71952712/5e331670-c098-422a-8309-92e58beadfc6">
Thanks for your patience!

**Comment by ChengkaiYang** [comment]
Sorry, i didn't notice that there has been"with torch.no_grad()".This question is finished!

**Comment by liuyueChang** [comment]
There is a mistake here,  p_old(z | y,x1:t) = p_old(z | x1:t) * p_old(y | z,x1:t) / **p_old(y | x1:t)**

I have a question about training objective

The loss calculation in the paper includes log p(Z | X) and log p(Y|Z,X), but in the code

[AutoBots/utils/train_helpers.py](https://github.com/roggirg/AutoBots/blob/3a61ad9f80603f99ab1e9cc535acccf72c5d6bea/utils/train_helpers.py#L72)

Line 72 in [3a61ad9](https://github.com/roggirg/AutoBots/commit/3a61ad9f80603f99ab1e9cc535acccf72c5d6bea)

 nll_k = nll_pytorch_dist(pred[kk].transpose(0, 1), data, rtn_loss=True) * post_pr[:, kk] 

only the p(Y|Z,X) participates in the loss calculation.

Please give me some tips, thank you!

**Comment by ChengkaiYang** [comment]
> There is a mistake here, p_old(z | y,x1:t) = p_old(z | x1:t) * p_old(y | z,x1:t) / **p_old(y | x1:t)**
> 
> I have a question about training objective
> 
> The loss calculation in the paper includes log p(Z | X) and log p(Y|Z,X), but in the code
> 
> [AutoBots/utils/train_helpers.py](https://github.com/roggirg/AutoBots/blob/3a61ad9f80603f99ab1e9cc535acccf72c5d6bea/utils/train_helpers.py#L72)
> 
> Line 72 in [3a61ad9](https://github.com/roggirg/AutoBots/commit/3a61ad9f80603f99ab1e9cc535acccf72c5d6bea)
> 
> nll_k = nll_pytorch_dist(pred[kk].transpose(0, 1), data, rtn_loss=True) * post_pr[:, kk]
> 
> only the p(Y|Z,X) participates in the loss calculation.
> 
> Please give me some tips, thank you!
log p(z| x) participates in the loss calculation as a part of KL_loss KL(p_old(z| y, X_1:t) || p(z|x)) but not reconstruction loss. So the CVAE loss has two parts, minimize reconstruction loss (minimize negative loglike loss) and minimize KL loss. Hope can help you.
